pubsub/tests/unit/subscriber_test.py

```diff
@@ -269,6 +269,38 @@ async def f(*args, **kwargs):
         assert producer_task.done()
         assert producer_task.exception()
 
+    @pytest.mark.asyncio
+    async def test_producer_gracefully_shutdowns(subscriber_client):
+        mock = MagicMock()
+        def g():
```

Comment:
Interesting 😄 I just want to point out that I think setting up `g` here as a generator doesn't quite represent the actual behavior we would see. For example, the `UnboundLocalError` case I mentioned above doesn't occur when it maybe should. 

You could also do something like

```python
async def f():
    raise asyncio.CancelledError
    
subscriber_clent.pull = f
```

Reply:
Oh wait this test is outright incorrect. The first message is already published by the time generator throws an exception 🤔

Reply:
Should be better now!